rfcs/20180817-variables-20.md

```diff
@@ -0,0 +1,140 @@
+# Variables in TensorFlow 2.0
+
+| Status        | Proposed       |
+:-------------- |:---------------------------------------------------- |
+| **Author(s)** | apassos@google.com |
+| **Sponsor**   | wicke@google.com, joshl@google.com, ashankar@google.com                 |
+| **Updated**   | 2018-08-17                                           |
+
+
+## Objective
+
+The API for TensorFlow variables has many drawbacks: impossible-to-reason-about semantics, reliance on global scopes, and reliance on global collections. As the TensorFlow API moves to become more pythonic and object oriented, with the Keras layers and models and the object-based serialization, we no longer have a need for much of this global infrastructure around variables.
+
+
+## Main changes
+
+The API for Variables will then change in the following ways for TF 2.0:
+
+
+
+*   tf.Variable will become an abstract base class with a well-defined interface and a scoped factory to construct instances
+    *   users will be able to implement their own variable-like objects by subclassing tf.Variable and adding a scoped factory function to use those variables
+*   variable_scope and get_variable will be removed
+    *   the tf 1.0 version of variable_scope and get_variable will be left in tf.compat.v1
+    *   to control variable naming users can use tf.name_scope + tf.Variable
+    *   whether a variable is shared across sessions / processes will be controlled by a constructor argument to tf.Variable; no other type of scope reuse will be done in the framework
+    *   scoped partitioning will be implemented as a factory function at first
+    *   libraries and users are encouraged to reuse variables by reusing their objects, like Keras layers do
+    *   custom_getters will have the following API: [variable_creator_scope](https://github.com/tensorflow/tensorflow/blob/567189980f7a1c2aa09a5170bd8d01a6ec37d303/tensorflow/python/ops/variable_scope.py#L2402)
```

Comment:
instead of relying on a scope, can we ask the variable factory explicitly?

Reply:
Why would you? tf.Variable is the only public API symbol for constructing variables, and it'll eventually bottom out to one or more calls to the base variable class ResourceVariable (RefVariable exists only in tf.compat.v1).

If you want to control what variables a piece of code can construct you can add a creator to the stack around that piece of code, but you should not have access to the lower bits because that would allow you to break the behavior of things like distribution strategies or make_template.

Can you clarify what are you trying to do?